cmd/ui.go

```diff
@@ -185,35 +252,33 @@ func showProgress(ctx context.Context, conf Config, execScheduler *local.Executi
 	//description in the TODO message in cmd/root.go)
 	if stdoutTTY && !noColor {
 		updateFreq = 100 * time.Millisecond
-		outMutex.Lock()
-		stdout.PersistentText = progressBarsPrint
-		stderr.PersistentText = progressBarsPrint
-		outMutex.Unlock()
-		defer func() {
-			outMutex.Lock()
-			stdout.PersistentText = nil
-			stderr.PersistentText = nil
-			if ctx.Err() != nil {
-				// Render a last plain-text progressbar in an error
-				progressBarsLastRender = []byte(renderMultipleBars(stdoutTTY, false, leftMax, pbs))
-				progressBarsPrint()
-			}
-			outMutex.Unlock()
-		}()
 	}
 
 	ctxDone := ctx.Done()
 	ticker := time.NewTicker(updateFreq)
+	sigwinch := NotifyWindowResize()
+	fd := int(os.Stdout.Fd())
 	for {
 		select {
-		case <-ticker.C:
-			barText := renderMultipleBars(stdoutTTY, true, leftMax, pbs)
-			outMutex.Lock()
-			progressBarsLastRender = []byte(barText)
-			progressBarsPrint()
-			outMutex.Unlock()
 		case <-ctxDone:
+			// FIXME: Remove this...
+			// Add a small delay to allow executors time to process
+			// the done context, so that the correct status symbol is
+			// outputted for each progress bar.
+			time.Sleep(50 * time.Millisecond)
```

Comment:
Any ideas for handling this... gracefully? :D This seemed like a quick fix for an annoying issue.

Reply:
No, it's only a problem with `^C`.

Reply:
I have to dig in deeper, but I think that means we should probably address the cancellation chain after `Ctrl+C`. Maybe something like this - Ctrl+C aborts the Engine context, but not the one passed to the progressbar rendering code - it is aborted only after the engine returns, either normally or when interrupted?

Reply:
That makes sense. I'll take a look at how that could be accomplished.

Related to `^C`, what do you think about [this comment](https://github.com/loadimpact/k6/blob/92a9716e6b3ac3574cd9a449e0afdad584c75b5a/cmd/run.go#L348-L350) now that #971 is fixed?

I like the idea that the first `^C` would trigger a "soft" shutdown ("SIGTERM", running something like `gracefulStop()` to allow all iterations to complete), but a second `^C` would interrupt that and stop immediately ("SIGKILL"). Could we plan this as part of #1279?

Reply:
I no longer think we should wait for iterations to complete on a first Ctrl+C interruption - it could potentially take minutes, and would usually be at least several seconds. Also, it would add significant implementation complexity for a dubious benefit. I can't think of a valid use case of why someone would care if an iteration doesn't complete when they have interrupted k6 with Ctrl+C...

In fixing the Ctrl+C handling, beyond the progressbar fix I mentioned above, I think we should modify the code like this:
1. The first Ctrl+C does what it currently does - interrupts the context for the `Engine`/`ExecutionScheduler` and waits for them to finish. Mind you, this is still graceful in a sense, we're interrupting the k6 execution in a very orderly manner and nothing is lost.
2. The second Ctrl+C directly calls [`os.Exit()`](https://golang.org/pkg/os/#Exit) with some [predefined exit code](https://github.com/loadimpact/k6/blob/9fb6466409db9ae6ca59617d95bcd5c25fa00169/cmd/run.go#L54-L60). This would prevent anything like https://github.com/loadimpact/k6/issues/971 happening again. Or, rather, if something like that happens, it's not going to be a big issue.

Reply:
So 1d28f2b4 unintentionally fixed this issue altogether! :tada:

No need for hacks and special Ctrl+C handling, all progress bar status symbols are rendered correctly now. I'm puzzled as to why that is exactly, but didn't look further into it.

So I don't think we need this refactor anymore. :) Test from the latest push and let me know.